R/checkout.R

```diff
@@ -28,64 +55,142 @@ checkout <- function(repos = getOption("repos"),
   renv_project_lock(project = project)
 
   # set new repositories
+  repos <- repos %??% renv_checkout_repos(date)
   options(repos = repos)
 
+  # TODO: Activate Bioconductor if it appears to be used by this project
+
   # select packages to install
-  packages <- packages %||% renv_checkout_packages(project = project)
-  names(packages) <- packages
+  packages <- packages %??% renv_checkout_packages(project = project)
+
+  # get the associated remotes for these packages
+  remotes <- renv_checkout_remotes(packages, project)
+
+  # parse these into package records
+  records <- map(remotes, renv_remotes_resolve)
+
+  # create a lockfile matching this request
+  lockfile <- renv_lockfile_init(project)
+  lockfile$Packages <- records
+
+  # restore from that lockfile
+  restore(lockfile = lockfile, clean = clean)
+
+}
+
+renv_checkout_packages <- function(project) {
```

Comment:
This surprised me — I had expected that you take the package names from the existing lock file.

Reply:
I don't think we want to use the lockfile -- the dependency tree is likely to change across different snapshots (e.g. IIRC `dplyr` only recently-ish took a dependency on `vctrs`) so we basically need to re-compute the dependency tree based on the packages used in the project.

In a sense, `renv::checkout()` is really like `renv::init()`, except instead of copying packages from the user / default library paths, we're copying them from a package repository.

Reply:
Oh, I hadn't thought about the dependency problem. Thanks for explaining!